Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/signature/core/DscSignatureValidator.kt

```diff
@@ -2,16 +2,144 @@ package de.rki.coronawarnapp.covidcertificate.signature.core
+
+    private val oidSet = vcOids + tcOids + rcOids
+
+    /**
+     * @throws InvalidHealthCertificateException if validation fail, otherwise it is OK!
+     */
+    suspend fun validateSignature(dccData: DccData<*>) {
+        val dscData = dscRepository.dscData.first()
+        Timber.tag(TAG).d("validateSignature(dscListSize=%s)", dscData.dscList.size)
+
+        findDscCertificate(dscData, dccData.dscMessage).apply {
+            validate()
+            checkOidsIntersect(dccData)
+        }
+    }
+
+    private fun findDscCertificate(
+        dscData: DscData,
+        dscMessage: DscMessage
+    ): X509Certificate {
+        val toVerify = dscMessage.signedPayload()
+        val filteredDscSet = dscData.dscList.filter { it.kid == dscMessage.kid }
+        Timber.d("filteredDscSetSize=${filteredDscSet.size}")
+
+        val matchedDscSet = when {
+            filteredDscSet.isEmpty() || dscMessage.kid.isEmpty() -> dscData.dscList
+            else -> filteredDscSet
+        }
+        Timber.d("matchedDscSetSize=${matchedDscSet.size}")
+
+        var x509Certificate: X509Certificate? = null
+        for (dsc in matchedDscSet) {
+            val dscCertificate = dsc.toX509certificate()
+            val (publicKey, signature) = when (dscMessage.algorithm) {
+                ES256 -> dscCertificate.publicKey to dscMessage.signature.toByteArray().toECDSAVerifier()
+                PS256 -> dscCertificate.publicKey.toRsaPublicKey() to dscMessage.signature.toByteArray()
+            }
+
+            try {
+                val valid = Signature.getInstance(dscMessage.algorithm.algName).verify(publicKey, toVerify, signature)
+                Timber.d("Dsc certificate (${dsc.kid}) is valid=$valid")
+
+                if (valid) {
+                    x509Certificate = dscCertificate
+                    break
+                }
+            } catch (ignored: Exception) {
+                // Ignore errors / continue
+            }
+        }
+
+        return x509Certificate ?: throw InvalidHealthCertificateException(HC_DSC_NO_MATCH)
+    }
+
+    private fun DscItem.toX509certificate(): X509Certificate = CertificateFactory
+        .getInstance("X.509")
+        .generateCertificate(data.toByteArray().inputStream())
+        as X509Certificate
+
+    private fun X509Certificate.validate() {
+        try {
+            checkValidity()
+        } catch (e: CertificateExpiredException) {
+            throw InvalidHealthCertificateException(HC_DSC_EXPIRED)
+        } catch (e: CertificateNotYetValidException) {
+            throw InvalidHealthCertificateException(HC_DSC_NOT_YET_VALID)
+        }
+    }
+
+    private fun X509Certificate.checkOidsIntersect(dccData: DccData<*>) {
+        val extendedKeysIntersect = extendedKeyUsage.orEmpty().toSet() intersect oidSet
+        if (extendedKeysIntersect.isEmpty()) return // OK!
+        when (dccData.certificate) {
+            is VaccinationDccV1 -> if (vcOids.intersect(extendedKeysIntersect).isEmpty())
```

Comment:
Would prefer to have brackets on multiline `if`s.

Reply:
brakets will make it noisy , both are valid syntax for one  `if`   statement I prefer it that way

Reply:
Hm, subjective.

Not a blocker for me.